pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,523 @@
+			}
+		}
+
+	}
+
+	// Delete chains no longer in use:
+	activeChains := make(map[utiliptables.Chain]bool) // use a map as a set
+	for _, chain := range newHostChains {
+		activeChains[chain] = true
+	}
+	for _, chain := range newServiceChains {
+		activeChains[chain] = true
+	}
+	for chain := range existingChains {
+		if !activeChains[chain] {
+			chainString := string(chain)
+			// Ignore chains that aren't from services.
+			// Short-circuit fail on chains without the shared prefix of services and general services chain,
+			// then fail more specifically.
+			if !strings.HasPrefix(chainString, "KUBE-S") || !(strings.HasPrefix(chainString, "KUBE-SVC-") || strings.HasPrefix(chainString, "KUBE-SERVICES")) {
+				continue
+			}
+			rulesEnd.WriteString(fmt.Sprintf("-X %s\n", chain))
+		}
+	}
+
+	// write end of table
+	rulesEnd.WriteString("COMMIT\n")
+	// combine parts
+	lines := append(rules.Bytes(), rulesEnd.Bytes()...)
+
+	// sync rules and return error
+	glog.V(3).Infof("Syncing rule: %s", lines)
+	// NOTE: flush=false is used so we don't flush all rules.
+	// we manually flush the rules we need to flush.
+	// counters=true is so we can preserve the counters
+	err = proxier.iptables.Restore(utiliptables.TableNAT, lines, utiliptables.NoFlushTables, utiliptables.RestoreCounters)
+	return err
+}
+
+// return an iptables-save/restore formatted chain line given a Chain
+func makeChainLine(chain utiliptables.Chain) string {
+	return fmt.Sprintf(":%s - [0:0]\n", chain)
+}
+
+// getChainLines parses a table's iptables-save data to find chains in the table.
+// It returns a map of iptables.Chain to string where the string is the chain line from the save (with counters etc).
+func getChainLines(table utiliptables.Table, save []byte) map[utiliptables.Chain]string {
+	// get lines
+	lines := strings.Split(string(save), "\n")
+	chainsMap := make(map[utiliptables.Chain]string)
+	tablePrefix := "*" + string(table)
+	lineNum := 0
+	// find beginning of table
+	for ; lineNum < len(lines); lineNum++ {
+		if strings.HasPrefix(strings.TrimSpace(lines[lineNum]), tablePrefix) {
+			lineNum++
+			break
+		}
+	}
+	// parse table lines
+	for ; lineNum < len(lines); lineNum++ {
+		line := strings.TrimSpace(lines[lineNum])
+		if strings.HasPrefix(line, "COMMIT") || strings.HasPrefix(line, "*") {
+			break
+		} else if len(line) == 0 || strings.HasPrefix(line, "#") {
+			continue
+		} else if strings.HasPrefix(line, ":") && len(line) > 1 {
+			chain := utiliptables.Chain(strings.SplitN(line[1:], " ", 2)[0])
+			chainsMap[chain] = lines[lineNum]
+		}
+	}
+	return chainsMap
+}
+
+// Assumes proxier.mu is held
+func (proxier *Proxier) removeServiceInternal(name proxy.ServicePortName, info *serviceInfo) {
+	delete(proxier.serviceMap, name)
+}
+
+// assumes proxier.mu is held
+func (proxier *Proxier) addServiceInternal(service proxy.ServicePortName) *serviceInfo {
```

Comment:
make call-sites more obvious - rename this to newServiceInfo() and make the callers do the insert.  Then it doesn't have to be a method.  And it can sit next to struct serviceInfo.